Reland: [clang] Diagnose dangling issues for the "Container<GSLPointer>" case. #107213

hokein opened the pull request:
This relands #107213, with with fixes to address false positives (`make_optional(nullptr)`).

hokein committed:
Reapply "[clang] Diagnose dangling issues for the "Container<GSLPointer>" case. (#107213)"

This reverts commit 0683c4e839524c37fe4ddfa1bce1e31ba556041b.

hokein committed:
Fix the false positive for make_optional(nullptr);

We should also consider isNullPtrType as the pointer type.

Xazax-hun reviewed (approved):
LGTM!

hokein committed:
fix another false positive caused by the interaction between the gsl::Owner and lifetimebound attributes.

usx95 commented on clang/test/Sema/warn-lifetime-analysis-nocfg.cpp:
```diff
@@ -633,4 +633,12 @@ std::optional<int*> test4(int a) {
   return std::make_optional(nullptr); // fine
 }
 
+template <typename T>
+struct [[gsl::Owner]] StatusOr {
+  const T &value() [[clang::lifetimebound]];
+};
```
Some ideas: https://godbolt.org/z/Y6nPM9j1b
Feel free to expand and add more. 
```cpp
#include<string>
#include <vector>

template <typename T>
struct [[gsl::Owner]] StatusOr {
  const T &value() const [[clang::lifetimebound]];
};

template<typename T>
struct [[gsl::Pointer]] Span {
    Span(const std::vector<T> &V): V(V){}
    const std::vector<T> &V;

    const int& getField() [[clang::lifetimebound]] { return field;}
    int field;
};

/////// From Owner<Pointer> ///////

// Pointer from Owner<Pointer>
std::string_view test6(StatusOr<std::string_view> aa) {
  return aa.value(); // Ok.
}

// Pointer<Owner<Pointer>>> from Owner<Pointer>
Span<int*> test9(std::vector<int*> V) {
  return V; // No error. !!FIXME: GH108463!!
}

/////// From Owner<Owner<Pointer>> ///////

// Pointer from Owner<Owner<Pointer>>
int* test8(StatusOr<StatusOr<int*>> aa) {
  return aa.value().value(); // Ok.
}
// Owner<Pointer> from Owner<Owner<Pointer>>
std::vector<int*> test5(StatusOr<std::vector<int*>> aa) {
  return aa.value(); // Ok.
}

// Pointer<Owner<Pointer>>> from Owner<Owner<Pointer>>
Span<int*> test8(StatusOr<std::vector<int*>> aa) {
  return aa.value(); // Error. Ok.
}

/////// From Owner<Owner> ///////

// Pointer<Owner>> from Owner<Owner>
Span<int> test9(StatusOr<std::vector<int>> aa) {
  return aa.value(); // Error. Ok.
}

/////// From Owner<Pointer<Owner>> ///////

// Pointer<Owner>> from Owner<Pointer<Owner>>
Span<int> test9(StatusOr<Span<int>> aa) {
  return aa.value(); // Ok.
}

// Lifetimebound and gsl::Pointer.
const int& test10(Span<int> a) {
  return a.getField();
}
```

hokein replied:
Thanks a lot for these tests. I've added them with some adjustments.

usx95 reviewed (commented):
We seem to be good at detecting new false positives but this is natural due to the visible compiler diagnositc. Unfortunately, same is not true for new false-negatives. More tests in our test-suite is the only way to detect those and we should be extensively adding more tests. This is something I feel particularly missing for lifetime analysis given the complexity of the code. This is obviously not actionable and non-blocking and is more of a meta comment.

See my comment for ideas for more tests.

I also found a recently introduced false-negative. This is not particularly introduced by this PR but is very related:
https://github.com/llvm/llvm-project/issues/108463

hokein committed:
Fix a newly-introcuded false negative.

And add more tests per the comment. The new false negative was caused by
replacing the instances of "!isRecordWithAttr<OwnerAttr>" in `TemporaryVisitor`
with "isGSLOwner".

hokein commented on clang/lib/Sema/CheckExprLifetime.cpp:
```diff
@@ -347,6 +361,30 @@ static bool shouldTrackFirstArgument(const FunctionDecl *FD) {
+      !isContainerOfPointer(ClassD))
+    return false;
+  return isGSLOwner(FirstArgType) ||
+         isStdInitializerListOfPointer(FirstArgType->getAsRecordDecl());
```
> It looks like we mean to say that std::initialiser_list<Pointer> should be considered as GSL Owner.

I think what we actually intend is to continue tracking the argument of type `std::initializer_list<Pointer>`. (while previously, we only tracked arguments for GSL Owner types.) Conceptually, `std::initializer_list<Pointer>` is not an owner, so moving this logic to `isGSLOwner` doesn't feel appropriate, even though it works in practice.

Regarding the new false negative, it was introduced when we replaced instances of isRecordWithAttr<OwnerAttr> in [TemporaryVisitor](https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/CheckExprLifetime.cpp#L988) with the new isGSLOwner function (which excludes `Owner<Pointer>` types). The logic in `TemporaryVisitor` is a bit subtle, which is to skip reporting issues if the local variable is not an owner type. Since `isGSLOwner` excludes `vector<int*>`, it no longer triggers a report.

To fix this, I reverted the abstraction of isGSLOwner because the `!isContainerOfPointer` logic only applies to determining whether to trace the first argument of the `container<pointer>` constructor. I've added comments there).

usx95 replied:
Thanks for the clarification.

hokein commented:
> We seem to be good at detecting new false positives but this is natural due to the visible compiler diagnositc. Unfortunately, same is not true for new false-negatives. More tests in our test-suite is the only way to detect those and we should be extensively adding more tests. This is something I feel particularly missing for lifetime analysis given the complexity of the code. This is obviously not actionable and non-blocking and is more of a meta comment.

+1, I fully agree with your point. I’ve felt that the current test lacks sufficient coverage. On the other side, I think the situation has been getter better with our recent improvements to lifetime analysis. Hopefully, with continued effort, we’ll reach a point where the test cover the majority of key cases.

> See my comment for ideas for more tests.
> 
> I also found a recently introduced false-negative. This is not particularly introduced by this PR but is very related: #108463

Thanks, please see my replied comment.

usx95 reviewed (approved):
LGTM. Thanks.

hokein committed:
Refine the heuristic on inferring the nested pointer onwnership.

Fix another new false positive

usx95 commented on clang/lib/Sema/CheckExprLifetime.cpp:
```diff
@@ -357,35 +369,101 @@ static bool shouldTrackFirstArgument(const FunctionDecl *FD) {
+  if (!ParamRefType)
+    return false;
+
+  // Check if the first parameter type "Owner<U>".
```
nit: `// Check if the first parameter type **is** "Owner<U>".`

hokein replied:
Done.

usx95 commented on clang/lib/Sema/CheckExprLifetime.cpp:
Can you also add a test for this with a FIXME.

hokein replied:
Added.

hokein committed:
Address review comments.

hokein commented on clang/lib/Sema/CheckExprLifetime.cpp:
```diff
@@ -285,6 +285,18 @@ static bool isContainerOfPointer(const RecordDecl *Container) {
 }
+static bool isContainerOfOwner(const RecordDecl *Container) {
+  if (const auto *CTSD =
+          dyn_cast_if_present<ClassTemplateSpecializationDecl>(Container)) {
```
Done.
